validation: Templatize ValidationState instead of subclassing

jkczyz opened the pull request:
This removes boilerplate code in the subclasses which otherwise only
differ by the result type.

The subclassing was introduced in a27a295.

ryanofsky reviewed (approved):
Code review ACK aa67e6d736fab1cf125687f3393a3eb259fb8608.

Medium concept ACK. Nice that this eliminates duplicate code. Added #includes and template function might slow down compilation a bit, though.

promag commented on src/consensus/validation.h:
```diff
@@ -71,31 +71,32 @@ enum class BlockValidationResult {
-protected:
-    void Invalid(const std::string &reject_reason="",
+public:
+    ValidationState() : m_mode(MODE_VALID) {}
```
nit, in L85 do `} m_mode{MODE_VALID};` and drop constructor?

jkczyz replied:
The diff here is a bit misleading as I'm actually moving `Invalid` to the `public` section and not modifying the constructor. Is the convention to put this in a separate a commit? Or would it be considered outside the scope of the PR?

promag replied:
Right, but the constructor could still be dropped.

jkczyz replied:
Understood. But that doesn't answer either of my questions... 🤔

promag replied:
IMHO could be added in "Templatize ValidationState instead of subclassing", but no big deal.

jkczyz replied:
Ended up doing this separately in 7470037 as it gives cleaner diffs.

promag commented on src/consensus/tx_check.h:
```diff
@@ -5,6 +5,8 @@
 #ifndef BITCOIN_CONSENSUS_TX_CHECK_H
 #define BITCOIN_CONSENSUS_TX_CHECK_H
 
+#include <consensus/validation.h>
```
You can still forward declare `TxValidationState` and `BlockValidationState` where needed and avoid the issues pointed by @ryanofsky?

jkczyz replied:
This causes a compilation error:

```console
In file included from consensus/tx_check.cpp:8:
./consensus/validation.h:115:7: error: typedef redefinition with different types ('ValidationState<TxValidationResult>' vs 'TxValidationState')
using TxValidationState = ValidationState<TxValidationResult>;
      ^
./consensus/tx_check.h:16:7: note: previous definition is here
class TxValidationState;
```

I'd have to forward declare the full template type and enum then use that in the function declaration.

```c++
enum class TxValidationResult;
template <typename Result>
class ValidationState;

bool CheckTransaction(const CTransaction& tx, ValidationState<TxValidationResult>& state);
```

Seems a bit verbose and would mismatch the function definition. FWIW, here are some [pros and cons](https://google.github.io/styleguide/cppguide.html#Forward_Declarations) of forward declaring that I'm aware of.

promag reviewed (commented):
Concept ACK.

laanwj commented:
> Added #includes and template function might slow down compilation a bit, though.

Yes, although this is a quite straightforward use of templates, no boost nested template horrors, so I hope it's not too bad.

ajtowns commented on src/util/validation.h:
```diff
@@ -6,12 +6,18 @@
 /** Convert ValidationState to a human-readable message for logging */
-std::string FormatStateMessage(const ValidationState &state);
+template <typename Result>
+std::string FormatStateMessage(const ValidationState<Result>& state)
```
Could move this into ValidationState directly, and make it
```
std::string FormatStateMessage() const { return m_reject_reason + (m_debug_message.empty() ? "" : ", " + m_debug_message); }
```

jkczyz replied:
Done in 7239d01 but named it `ToString` since #16688 uses it for valid states as well.

ajtowns commented on src/consensus/validation.h:
```diff
@@ -106,40 +107,13 @@ class ValidationState {
-    }
-    BlockValidationResult GetResult() const { return m_result; }
-};
+using TxValidationState = ValidationState<TxValidationResult>;
```
I think you can do `class TxValidationState : ValidationState<TxValidationResult> { };` and still be able to use forward declarations `class TxValidationState;` fwiw.

jkczyz replied:
Good idea! Done in 0468628.

ajtowns reviewed (commented):
Concept ACK

jkczyz committed:
Templatize ValidationState instead of subclassing

This removes boilerplate code in the subclasses which otherwise only
differ by the result type.

jnewbery commented:
Concept ACK. Will review code once @ajtowns's comments are addressed.

jkczyz committed:
Refactor FormatStateMessage into ValidationState

promag reviewed (commented):
Nice suggestions from @ajtowns!

practicalswift commented:
Concept ACK: new code is easier to reason about

jnewbery commented on src/util/validation.cpp:
```diff
@@ -5,15 +5,4 @@
-        state.GetDebugMessage().empty() ? "" : ", "+state.GetDebugMessage());
-}
-
 const std::string strMessageMagic = "Bitcoin Signed Message:\n";
```
It seems a bit sad to leave this const all by itself in this implementation file. Perhaps move it to the header file and change to a constexpr? Bonus points for renaming `util/validation.h` to `util/strings.h` since it isn't anything to do with validation any more, and I can't find anywhere that would be a better home for this const.

MarcoFalke replied:
I don't thing `std::string` can be constexpr. Also, moving it to the header file might inline it in every translation unit. Probably doesn't matter for a short string like this, but no need to change existing code to make it worse.

jnewbery replied:
You're right. `std::string` can't be constexpr.

jkczyz replied:
Looking at the references to `strMessageMagic`, they are always of the form:

```c++
CHashWriter ss(SER_GETHASH, 0);
ss << strMessageMagic;
ss << strMessage;
```

And then the result of its `GetHash` is passed to either `SignCompact` or `RecoverCompact`, and the variable `ss` is never used again.

So the above code could be refactored into a function where `strMessageMagic` is just an implementation detail. Given that, is there a more appropriate place for it? What does `ss` stand for here?

jnewbery replied:
> What does ss stand for here?

Not sure. Serialization stream?

jnewbery replied:
(there have been `CDataStream ss` variables in the codebase since Satoshi era)

jkczyz replied:
Perhaps I could move that boilerplate code to `util/hash.h` as `HashMessage`?

```c++
uint256 HashMessage(const std::string& message);
```

Open to more suitable naming for the file and function. Note that `util/string.h` already exists for string joining, so the plural version may be confusing.

Would likely leave this to a follow-up PR given I'd be refactoring call sites.

jnewbery replied:
> Open to more suitable naming for the file and function

Perhaps `/util/message.h`?

> Would likely leave this to a follow-up PR given I'd be refactoring call sites.

Yeah, anything other than a simple rename should be left as a follow-up PR.

jkczyz replied:
Opened #17557 with the refactor.

jkczyz replied:
Rebased and was able to remove `util/validation.h` and `util/validation.cpp` now that #17577 is merged.

jnewbery reviewed (commented):
Tested ACK 0468628e8191c574d2568f5e9e7fa15dd43d80f2. Nice tidy up!

One suggestion for getting rid of a now almost-empty file. Can be done in this PR or a follow-up.

promag reviewed (commented):
Code review ACK 0468628e8191c574d2568f5e9e7fa15dd43d80f2.

Restarted failed job on travis.

jkczyz committed:
Remove ValidationState's constructor

jnewbery commented:
utACK 2948ebd29d0a53f11aefc8546bcac4125c726cc6

jkczyz commented:
Rebased and updated to account for #17624 in 1452949.

ariard reviewed (commented):
Code review ACK 1452949.

About the increase compilation time concern, on my system compare to master it's even a bit slightly slightly faster.

MarcoFalke commented:
ACK 14529499743e071ddc091e5508e1eea4532e754d 🎏

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 14529499743e071ddc091e5508e1eea4532e754d 🎏
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgCpgv/XrM9w3dqV6IqX53LVCU81lx6hex2Gjbb8X0wb2rfaBxS1H/y+DHq9cPs
iBh6bp5HGABL2dRPcS3mvyBzA9lbWu4MZVzqu728m9m2tUdMjnezxzz0tZOhmcWL
vW8+UpYVnr4UahR9k5X2N06v7YoqkpNmQ06aSOkRCNd7XPkjOAh3SN5zOSUjs1n8
2wUOPPNAbBS5UgFt6vTPTzm6AMNWp+fXtqnSIe84JFkefEoFo0EW1f72Nyb1agrG
70Cxd+Ph4de0yZq+lvcrZz62JPOGwF8AWW8ls+82hOFMmmxI1eduQPo9xuEvv14b
euvSNwjmUenZBhldWXAR0ct13O3NNgSzdbWZjBomqtDFDa5MpcOGHeCppYmxykD7
7FX/+Ye7Q181D7vXBFj/YBEGqPyIuLIm0NV1N6EPl8gYiqaK0c7UF0TJPqwWE6k1
HjhDdT4cD9rZrcl9u6mt2zdjcavNuyfMO13seNINZVXuNkgleta/sCh2edQU41XL
I09hdKhh
=2AEE
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `70e15c1a24082f6b136c95c7920ed1bc271fccb39fdfee1f98976554082edd59  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010870e15c1a24082f6b136c95c7920ed1bc271fccb39fdfee1f98976554082edd59f010783e58a3a72f271f80bdb6f6a5de3be408fff0102b3bf6a50f02743fe474e6e8dab0814608f120d5b1a54068d415f56bcc0611903b53f0c2194774d9daa9af1cae5b9868bc9b1208f1045dfd0c20f008868906f2ae5f7d540083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010620c04ef6024ea366294178f9799881d08f120cef74e9de472802d3fcd812475c727cc476312cf7b159922beead31f4c7c6e0508f1045dfd0c21f0084b3c94d916e692320083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01064f363bb1c140cbe686e93afc96c067908f120866289f6f993be72149a008dfe90acf74a2c819383d6f72a4414d0f34c449f3508f1045dfd0c20f0089264421241f61ebf0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010acfdd9c50f9e11cbce48c33b2a1953c608f1045dfd0c21f0089c5cca81aa5140240083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

jnewbery commented:
utACK 14529499743e071ddc091e5508e1eea4532e754d

Since the initialization relies on `[TX|BLOCK]_RESULT_UNSET` being 0, what do you think of:

```diff
--- a/src/consensus/validation.h
+++ b/src/consensus/validation.h
@@ -16,7 +16,7 @@
   * provider of the transaction should be banned/ignored/disconnected/etc.
   */
 enum class TxValidationResult {
-    TX_RESULT_UNSET,         //!< initial value. Tx has not yet been rejected
+    TX_RESULT_UNSET = 0,     //!< initial value. Tx has not yet been rejected. Must be zero/first defined enumerator for default initialization.
     TX_CONSENSUS,            //!< invalid by consensus rules
     /**
      * Invalid by a change to consensus rules more recent than SegWit.
@@ -50,7 +50,7 @@ enum class TxValidationResult {
   * useful for some other use-cases.
   */
 enum class BlockValidationResult {
-    BLOCK_RESULT_UNSET,      //!< initial value. Block has not yet been rejected
+    BLOCK_RESULT_UNSET = 0,  //!< initial value. Block has not yet been rejected. Must be zero/first defined enumerator for default initialization.
     BLOCK_CONSENSUS,         //!< invalid by consensus rules (excluding any below reasons)
     /**
      * Invalid by a change to consensus rules more recent than SegWit.
```

zachwylde00 commented:
Good evening Gentleman,

Thank you for your reconsideration of #17399. I could not have requested
this at my error and I extend gratitude for validation. Truly, I hope to
not make such easily avoidable errors by utilization of optimization and
other implementations that great minds, before my maybe great mind, set as
guides. I have noted my errors and updated my checklist. Enjoy your evening
and once again, sorry for the err.



On Fri, Dec 20, 2019, 5:27 PM John Newbery <notifications@github.com> wrote:

> utACK 1452949
> <https://github.com/bitcoin/bitcoin/commit/14529499743e071ddc091e5508e1eea4532e754d>
>
> Since the initialization relies on [TX|BLOCK]_RESULT_UNSET being 0, what
> do you think of:
>
> --- a/src/consensus/validation.h+++ b/src/consensus/validation.h@@ -16,7 +16,7 @@
>    * provider of the transaction should be banned/ignored/disconnected/etc.
>    */
>  enum class TxValidationResult {-    TX_RESULT_UNSET,         //!< initial value. Tx has not yet been rejected+    TX_RESULT_UNSET = 0,     //!< initial value. Tx has not yet been rejected. Must be zero/first defined enumerator for default initialization.
>      TX_CONSENSUS,            //!< invalid by consensus rules
>      /**
>       * Invalid by a change to consensus rules more recent than SegWit.@@ -50,7 +50,7 @@ enum class TxValidationResult {
>    * useful for some other use-cases.
>    */
>  enum class BlockValidationResult {-    BLOCK_RESULT_UNSET,      //!< initial value. Block has not yet been rejected+    BLOCK_RESULT_UNSET = 0,  //!< initial value. Block has not yet been rejected. Must be zero/first defined enumerator for default initialization.
>      BLOCK_CONSENSUS,         //!< invalid by consensus rules (excluding any below reasons)
>      /**
>       * Invalid by a change to consensus rules more recent than SegWit.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/17399?email_source=notifications&email_token=AKLXJUQ2HB47EJUM3A4YA5LQZVIFJA5CNFSM4JKAAHR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHOOBQQ#issuecomment-568123586>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKLXJURT5J6Q43NM4GSORODQZVIFJANCNFSM4JKAAHRQ>
> .
>

jkczyz commented:
> utACK [1452949](https://github.com/bitcoin/bitcoin/commit/14529499743e071ddc091e5508e1eea4532e754d)
> 
> Since the initialization relies on `[TX|BLOCK]_RESULT_UNSET` being 0, what do you think of:
> 
> ```diff
> --- a/src/consensus/validation.h
> +++ b/src/consensus/validation.h
> @@ -16,7 +16,7 @@
>    * provider of the transaction should be banned/ignored/disconnected/etc.
>    */
>  enum class TxValidationResult {
> -    TX_RESULT_UNSET,         //!< initial value. Tx has not yet been rejected
> +    TX_RESULT_UNSET = 0,     //!< initial value. Tx has not yet been rejected. Must be zero/first defined enumerator for default initialization.
>      TX_CONSENSUS,            //!< invalid by consensus rules
>      /**
>       * Invalid by a change to consensus rules more recent than SegWit.
> @@ -50,7 +50,7 @@ enum class TxValidationResult {
>    * useful for some other use-cases.
>    */
>  enum class BlockValidationResult {
> -    BLOCK_RESULT_UNSET,      //!< initial value. Block has not yet been rejected
> +    BLOCK_RESULT_UNSET = 0,  //!< initial value. Block has not yet been rejected. Must be zero/first defined enumerator for default initialization.
>      BLOCK_CONSENSUS,         //!< invalid by consensus rules (excluding any below reasons)
>      /**
>       * Invalid by a change to consensus rules more recent than SegWit.
> ```

Good call. Pushed a975974.

@promag @ariard @MarcoFalke You can verify the change with `git range-diff abb30de63 145294997 a975974f4`.

jnewbery commented:
ACK a975974

jkczyz commented:
Rebased

jnewbery commented:
reACK 2afff87a88a3f539cece488f927386a6e106cdc2

ariard commented:
ACK 2afff87, only change since 1452949 is default initialization first member of `TxValidationResult`/`BlockValidationResult` enums to 0.

jkczyz commented:
Rebased to account for changes in #16688. Specifically, pushed 7bb194ba252191b531803ad2952c984c70d192da to:

- Account for changes to `FormatStateMessage` made in 72f3227c83810936e7a334304e5fd7c6dab8e91b and 6edebacb2191373e76d79a4972d6192300976096
- Update use of `FormatStateMessage` in f9abf4ab6d3d3e4d4b7e90723020b5422a141a6f

Can be verified with:
```
git range-diff 17e14ac..2afff87 e7f8450..bf5bc26
```

jnewbery commented:
code review ACK bf5bc260e914e2413bed4041e5089fdee03a344a

MarcoFalke commented:
ACK 10efc0487c442bccb0e4a9ac29452af1592a3cf2 🐱

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 10efc0487c442bccb0e4a9ac29452af1592a3cf2 🐱
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiSBQv9GTJ4FVCrg3wq2XM0X75nYvO3+OK32UmnzgXDJjk7Pghohc7v7mUhqpnc
hmwCVwvvs9r5PdRx2BYx/GgPi6UEQ0KSjZfDxKUYcA7lMYCcHlHUxoohNj0mssN+
hpuGhzhBEYyX1jHNFazpSMSlKVPvesXQRHgo4Kw4EyyI1EU1Mpdi/nPWpsY6Of0y
Gg49gHJVvMp08RmkTVLRBWF0KqUATRfUTgqVUr5bit3tKZZPacxaiXSE8W8WXRa5
Nb6e1WHAa8+jWqlahlCFJnw4UoyXSIHtijhBgVliCo7wX5Lj588VjouQeaXEIufZ
q823LmNj8QdOwnNnE9wMSKYX/kQaGB+R3KRDpH+s6zHub86fA0/oz3dkER4XWaIM
DZNcYPnJYNgWtZ9f99uIC87pHeQ3kFA+5/xroVzjPvFEGLv7euwlgZn0v2rb2dB5
vgmIwvEH/B8/pH3GvL65yGCVFrkVA5rAnNGViav0KJujkWXukc0fkC9to5urHnwB
oz7K+tzI
=g9Fg
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `02d3cb1c7845b1a605977e2f801af0118c57df49376b61c77970f1807528dacf  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010802d3cb1c7845b1a605977e2f801af0118c57df49376b61c77970f1807528dacff01067a33d2c4cb0653aecb4d9dd7603921f08fff0103cdf3445824b8b715a09744b81bf93b008f1045e596745f008d1588ffa98851d1c0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01045e965a8f7bc04ba6d376ccf435f6cb608f0206121b35208c3101b4f197e2823273361d54476e26ae802d1f8d70a62b9d86c0608f1045e596746f008a442a4c67bba150d0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010505ac15f67bc13cfbcf064a62e784a3c08f020ed3f5d4576c9e0b0a953d9b08253d48f8868bdd7eb32a1571838ae1581feafb608f1045e596746f00863836ceede09d64f0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df0108423e7a1320f6d1278a7e979a229fdc908f020684c1c616bbb137e88994755ce0b2d8f8620c0ae12c8e78e9d93f3999201a34f08f1045e596746f0086c83777a7a80be1e0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

practicalswift commented:
ACK 10e85d4adc9b7dbbda63e00195e0a962f51e4d2c -- patch looks correct

ajtowns commented:
ACK 10efc0487c442bccb0e4a9ac29452af1592a3cf2 -- looks good to me

jonatack commented:
ACK 10efc048 code review, build/tests green, nice cleanup

practicalswift commented:
post-merge ACK 10efc0487c442bccb0e4a9ac29452af1592a3cf2
